mods to support blacklisted android devices

Cy-Bo-Rg opened the pull request:
I've added again a few lines to support blacklisted android devices with black screen problem (appearing since rev68).
I've added a modified index.html called index_mobile.html which fixes
resize problems and refresh fix problems in black listed android devices on stock browser.
I've removed a variable from a demo which was declared but was not used. issue #6766

My device is a Galaxy Tab 2.0 GT-P3110 running Android 4.1.2 on a Power VR SGX-540 (blacklisted)
Almost all demos works perfectly with this mod. Without the lines added in the "src/Three.js" , the device shows a black screen.
With the lines , it shows the demos but with a wrong refresh rate (1 to 5 fps) and the windows doesn't properly resize.
With the modified index_mobile.html the window resizes correctly and the refresh rate is at its maximum possible (60+fps on some demos).
On the same device the Crosswalk Chromium webview crashes sometimes and has a lower FPS than with this fix; the same happens on Chrome (latest dev build) and Firefox (latest beta build).
To make a comparison, the "material/skin" demo runs at 2 fps on crosswalk while on android stock browser with this fix runs at 8 fps (6-5 while rotating the view).

Related issues  #2119 (unreasonably closed)
There are many related issues but I couldn't find all of them because they reside in other revisions now.

Here are some screens of the android stock browser running three.js rev71.
Without lines added in src/Three.js , current index.html.
![screenshot_2015-07-04-05-57-03](https://cloud.githubusercontent.com/assets/5455606/8506408/eec2d8dc-2212-11e5-9a54-13148a8f6dad.png)

With lines added in the build but without modified index. Doesn't properly resize , the frame rate is wrong and the image refreshes at 1 to 5 fps.
![screenshot_2015-07-04-05-58-39](https://cloud.githubusercontent.com/assets/5455606/8506409/f68ebd42-2212-11e5-9f83-f1c9e810851f.png)

With lines added in the build and with index_mobile.html : 100% playable.
![screenshot_2015-07-04-06-03-48](https://cloud.githubusercontent.com/assets/5455606/8506438/990e7854-2214-11e5-9f2e-a81a4b85896e.png)

This video shows the problem and the fix in action.
https://www.youtube.com/watch?v=P_i0uOPWBuI
These modifications won't alter the behavior on desktop,which remains the same.

I've highly tested this refresh workaround since rev66. It always works.
But I know for sure that it is present starting from rev54 because I can run helloenjoy samsung racer s with the refresh fix.

Cy-Bo-Rg committed:
mods to support blacklisted android devices

I've added again a few lines to support blacklisted android devices.
I've added a modified index.html called index_mobile.html which fixes
resize problems, black-screen problems and refresh fix problems in black
listed android devices.
I've removed a variable from a demo which was unused.

Cy-Bo-Rg commented:
This is a screenshot of rev72 dev with my modified index_mobile.html.
Resizes correctly and the refresh rate is good,the shadows are rendered correctly even though no controls are showing up.
![screenshot_2015-07-04-17-00-50](https://cloud.githubusercontent.com/assets/5455606/8508388/49fad960-226e-11e5-840b-8a9625fc29e2.png)

Cy-Bo-Rg commented:
The refresh problem is a webgl bug and I'm letting them know of it but the black screen and the resize problem are some three.js bugs which can be fixed by adding those lines of code which were present in previous versions and by modifying the index.html accordingly.

Cy-Bo-Rg commented:
A few words are much appreciated.

mrdoob commented:
Android's stock browser doesn't support requestAnimationFrame?

Cy-Bo-Rg commented:
Someone says so, but he's wrong.
It's supported,at least partially.
https://youtu.be/UIualrl8FNo

Cy-Bo-Rg commented:
This is three.js rev72 dev on Android 4.2.2 on Mali 400 with its stock browser.

Rev72dev without code and refresh fix
![screenshot_2015-07-05-10-46-19](https://cloud.githubusercontent.com/assets/5455606/8511282/b0f68186-230f-11e5-8583-2d84fc8cf5d7.png)

Rev72dev with code added and fix
![screenshot_2015-07-05-11-32-30](https://cloud.githubusercontent.com/assets/5455606/8511287/d5b12a94-230f-11e5-9924-7153beef8141.png)
Rev72dev with code added and fix
![screenshot_2015-07-05-10-48-16](https://cloud.githubusercontent.com/assets/5455606/8511293/1d836792-2310-11e5-97cc-9c1316d7fc2e.png)
Rev72dev with code added and fix
![screenshot_2015-07-05-11-34-03](https://cloud.githubusercontent.com/assets/5455606/8511294/25028980-2310-11e5-994d-6862d29ae234.png)
Rev72dev with code added and fix
![screenshot_2015-07-05-11-35-16](https://cloud.githubusercontent.com/assets/5455606/8511296/2b4df450-2310-11e5-892b-f92672079c2d.png)


For a comparison with my blacklisted device I've tested the latest chrome dev on this Mali 400 (which isn't blacklisted) versus stock android browser on my Power VR SGX540.

Mali400 running "parallaxmap" demo on rev72dev
![screenshot_2015-07-05-11-57-21](https://cloud.githubusercontent.com/assets/5455606/8511305/557f4b48-2310-11e5-8081-a405bb8b8de8.png)
Mali400 running "parallaxmap" demo on rev72dev
![screenshot_2015-07-05-11-57-50](https://cloud.githubusercontent.com/assets/5455606/8511309/5dde5a18-2310-11e5-90f4-efd89c217fc8.png)

Power VR SGX540 running "parallaxmap" on rev72dev (modified)
![screenshot_2015-07-05-12-10-50](https://cloud.githubusercontent.com/assets/5455606/8511311/6d2fefe0-2310-11e5-8bc4-d4d26f02ef17.png)

Cy-Bo-Rg commented:
The last word is up to you guys.

Cy-Bo-Rg commented:
The devices which may benefit from such enhancements are 1497.
(http://www.gsmarena.com/results.php3?sOSes=2&sOSversions=2410

Jelly bean users (which can't upgrade or who don't care about changing browser to surf the web) are the vast majority.
!(http://blogs-images.forbes.com/ewanspence/files/2015/01/lollipop-dead-graph.jpg)

Cy-Bo-Rg commented:
"Android's stock browser doesn't support requestAnimationFrame?"
Xperia MT27i Pepper running andorid 2.3.7 on Mali 400 ,stock browser 2.3.7.
I'm sorry but this phone cannot even take screenshots without root access.
![20150705_192448](https://cloud.githubusercontent.com/assets/5455606/8512638/0d33f964-234d-11e5-93ee-314ab9bcce0e.jpg)
Current index.html with rev72dev without added code doesn't even show up (if I'm not wrong,an html tag isn't closed somewhere...)
![20150705_190306](https://cloud.githubusercontent.com/assets/5455606/8512641/1f077cd8-234d-11e5-9518-e40cd37a8868.jpg)
My corrected index_mobile.html , instead, does show up correctly and as expected.
![20150705_190413](https://cloud.githubusercontent.com/assets/5455606/8512648/6daeb3a6-234d-11e5-9f7b-38d3587d88e7.jpg)
Rev72dev without mod : **black screen**(on all demos,this was a canvas demo).
![20150705_191337](https://cloud.githubusercontent.com/assets/5455606/8512662/d01d4908-234d-11e5-8e30-6637d7f2accc.jpg)
Rev72dev modded.
![20150705_190416](https://cloud.githubusercontent.com/assets/5455606/8512652/8dfca51e-234d-11e5-9840-61b34faee16d.jpg)
Rev72dev modded.
![20150705_190810](https://cloud.githubusercontent.com/assets/5455606/8512660/b339add6-234d-11e5-9fb1-b1873e775bd4.jpg)
Rev72dev modded.
![20150705_190952](https://cloud.githubusercontent.com/assets/5455606/8512661/c3205f60-234d-11e5-99a9-1dfe017c8008.jpg)
Rev72dev modded.
![20150705_191609](https://cloud.githubusercontent.com/assets/5455606/8512666/e9e8038c-234d-11e5-9cdc-15ed67e8b30f.jpg)
Rev72dev modded.
![20150705_191654](https://cloud.githubusercontent.com/assets/5455606/8512673/199ef5c2-234e-11e5-96fd-1ddde4a60a86.jpg)
Rev72dev modded.
![20150705_191823](https://cloud.githubusercontent.com/assets/5455606/8512675/24cc2316-234e-11e5-8e84-9505d54cd38c.jpg)

What else can I say ?
I don't know if you need more evidence than this, but avoiding the conversation while talking with others...Just close the pull request , I'll understand...

WestLangley commented:
> I don't know if you need more evidence than this, but avoiding the conversation while talking with others...Just close the pull request , I'll understand...

@Cy-Bo-Rg Thank you for your contributions.

Please relax and be patient. @mrdoob will get around to this in due time. Remember, it will be on his schedule, not yours.

Cy-Bo-Rg commented:
Absolutely, I recognize the efforts , I'm in love with his work , that's probably why I care so much, but you need one minute out of 2 days to make a couple of comments :P

WestLangley commented:
> but you need one minute out of 2 days to make a couple of comments

That is not how life works, dude. : - )

Cy-Bo-Rg commented:
I get it... :P
But I apologize for my index_mobile.html though; it has the demo list of three.js master revision 71,not 72dev.

Cy-Bo-Rg commented:
Sorry if I add many comments, but to have a laugh together, watch this... X)

http://caniuse.com/#feat=webgl

I mean , WHY ??? X)

Cy-Bo-Rg commented:
please notice that the page "http://twofuckingdevelopers.com/2013/06/gsap-requestanimationframe-performance-on-android-browsers/" isn't available anymore and I really don't know why or what happened since two days ago it was visible and ready to be seen...
BUT, there is a snapshot in the Archive.org available to be seen...
http://web.archive.org/web/20150316151742/http://twofuckingdevelopers.com/2013/06/gsap-requestanimationframe-performance-on-android-browsers/

mrdoob commented:
Would it be possible to implement the fixes of `index_mobile.html` into `index.html`? I would prefer to just have one file. Adding the requestAnimationFrame shim in `Three.js` should be ok.

Cy-Bo-Rg commented:
Sure, I'm on it. I'll fix the indentation too and I'll update it with rev72 demos.
I might specify in a comment that the "refresh fix" has to be applied in the html of a demo for it to work "standalone" without the index because if you open a demo without it, it will not refresh properly unless the hidden marquee is present.
But I can even apply it to all the demos if requested to do so.
Waiting for updates.
P.s. : do I have to close this pull request and open another one,right ?

Cy-Bo-Rg committed:
Modifications to support blacklisted Android chipsets

Index.html updated with refresh fix.
Some Android stock browsers don't support some css properties and the
width resizes to the whole screen in landscape mode. It works perfectly
in portrait mode though.
src/Three.js updated with requestanimationframe shim.
index_mobile.html removed.

Cy-Bo-Rg commented:
updated now. sorry for the late.

Cy-Bo-Rg committed:
removed unnecessary lines and fixed a typo

removed unnecessary lines and fixed a typo

Cy-Bo-Rg commented:
p.s: Firefox beta 40.0 has problem rendering most of the demos and with Chrome, no matter how many flags I activate, I just can't use it.

Cy-Bo-Rg commented:
sorry , I've pressed the button for error :P

mrdoob commented on examples/index.html:
```diff
@@ -564,7 +589,7 @@
 		}
 
-		</script>
+		</script><div class="marquee"><p>refresh_fix_on_some_blacklisted_android_chipsets</p></div></article>
```
Hold on a second. Blacklisted... does it mean that you've disabled the blacklisting? If so, wouldn't it be better to just support Firefox? I suspect very few people disable blacklisting.

Cy-Bo-Rg replied:
I've not disabled blacklisting on stock browser, it renders whatever you throw at it.
Stock browser doesn't have blacklist.
That's the point. Hundreds of millions of users would benefit.

Cy-Bo-Rg commented:
The index has been modified in that way because some css properties like "calc(100%-310)" are unsupported on many browsers on mobile while "width="100%" in the tag is perfectly recognized.
This doesn't affect rendering on desktop browsers though , which are relying on the css and which are working perfectly.

mrdoob commented:
Ok, adding `requestAnimationFrame` sound good, but the css changes seem kind of hacky.

Cy-Bo-Rg commented:
@mrdoob , the css isn't a problem for me, I mean, without it the frame doesn't scale properly and I see this 
![screenshot_2015-07-04-05-58-39](https://cloud.githubusercontent.com/assets/5455606/8506409/f68ebd42-2212-11e5-9f83-f1c9e810851f.png)
but it's not a problem considering that it can be fixed easily.
Even the refresh fix can be left alone if you don't want to add it (even though I really think of it as a useful implementation to keep backward compatibility with old chipsets).
If you could tell me what to eliminate and what to keep I'll fix the pull request a.s.a.p. today.

Cy-Bo-Rg committed:
removed the "massive hack"

refresh fix removed.

Cy-Bo-Rg commented:
@mrdoob 
Pull request updated.
"massive hack" removed...
It was really good by the way. If for whatever reason webgl should stop refreshing,it could have prevented it from happening.
But the requestAnimationFrame part is what's really needed, so, I can live without it.

mrdoob commented:
Thanks!

Cy-Bo-Rg commented:
Thank you, without you this couldn't be possible :)
